lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -791,105 +813,105 @@ makeChange
-                (NE.zipWith (<>)
-                    . makeChangeForNonUserSpecifiedAsset outputMapsOrdered)
-                (TokenMap.empty <$ outputMapsOrdered)
-                nonUserSpecifiedAssets
-
-        let change :: NonEmpty TokenMap
-            change = NE.zipWith (<>)
-                changeForUserSpecifiedAssets
-                changeForNonUserSpecifiedAssets
-
-        (bundles, remainder) <-
-            maybe (Left $ changeError excessCoin change) Right $
-                excessCoin `subtractCoin` requiredCost
-                >>=
-                runStateT
-                    (sequence (StateT . assignCoin minCoinValueFor <$> change))
-
-        let changeForCoins :: NonEmpty TokenBundle
-            changeForCoins = TokenBundle.fromCoin
-                <$> makeChangeForCoin outputCoins remainder
-
-        pure (NE.zipWith (<>) bundles changeForCoins)
+    | otherwise =
+        maybe (Left changeError) Right $ do
+            adaAvailable <- excessCoin `subtractCoin` requiredCost
+            assignCoinsToChangeMaps adaAvailable minCoinFor $
+                NE.zip changeForAssets outputCoinsOrdered
   where
+    -- The following subtraction is safe, as we have already checked
+    -- that the total input value is greater than the total output
+    -- value:
+    excess :: TokenBundle
+    excess = totalInputValue `TokenBundle.unsafeSubtract` totalOutputValue
+
+    (excessCoin, excessAssets) = TokenBundle.toFlatList excess
+
+    nonUserSpecifiedAssets = Map.toList $
+        F.foldr discardUserSpecifiedAssets mempty inputBundles
+
+    -- Change for user-specified assets: assets that were present in the
+    -- original set of user-specified outputs ('outputsToCover').
+    changeForUserSpecifiedAssets :: NonEmpty TokenMap
+    changeForUserSpecifiedAssets = F.foldr
+        (NE.zipWith (<>)
+            . makeChangeForUserSpecifiedAsset outputMapsOrdered)
+        (TokenMap.empty <$ outputMapsOrdered)
+        excessAssets
+
+    -- Change for non-user-specified assets: assets that were not present
+    -- in the original set of user-specified outputs ('outputsToCover').
+    changeForNonUserSpecifiedAssets :: NonEmpty TokenMap
+    changeForNonUserSpecifiedAssets = F.foldr
+        (NE.zipWith (<>)
+            . makeChangeForNonUserSpecifiedAsset outputMapsOrdered)
+        (TokenMap.empty <$ outputMapsOrdered)
+        nonUserSpecifiedAssets
+
+    -- Change for all assets: both user-specified and non-user-specified.
+    changeForAssets :: NonEmpty TokenMap
+    changeForAssets = NE.zipWith (<>)
+        changeForUserSpecifiedAssets
+        changeForNonUserSpecifiedAssets
+
+    -- Change for all assets, with the minimum ada amounts added.
+    changeForAssetsWithMinimalCoins :: NonEmpty TokenBundle
+    changeForAssetsWithMinimalCoins =
+        assignMinimumCoin minCoinFor <$> changeForAssets
+
+    totalMinimumCoinForAssets =
+        F.foldMap TokenBundle.getCoin changeForAssetsWithMinimalCoins
```

Comment:
I find this a little confusing now to have these two declarations in the top-level `where` because they are only used for error reporting. That's why I originally put `totalMinCoinValue` under `changeError`. IMO, the nicest feature about `where` is that it can be nested to be of limited scope and local to where it is used, which helps a lot with readability. Having a top-level flat list of local bindings reduces readability.

Reply:
Hi @KtorZ Good point!

I originally moved this up to a higher scope as I wanted to reuse it from another place. However, I eventually changed my implementation so that this was no longer necessary.

I'm happy to move this back under the `where` clause (where it was before).

Reply:
Fixed in 3ca8552436fd1ccc23e1256f27e34032d7adbae6.